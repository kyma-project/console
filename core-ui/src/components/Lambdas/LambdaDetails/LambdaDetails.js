import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import LuigiClient from '@kyma-project/luigi-client';
import { useMutation } from '@apollo/react-hooks';
import { TabGroup, Tab } from 'fundamental-react';
import LabelSelectorInput from '../../LabelSelectorInput/LabelSelectorInput';

import { UPDATE_LAMBDA } from '../../../gql/mutations';
import LambdaDetailsHeader from './LambdaDetailsHeader/LambdaDetailsHeader';

import { useNotification } from 'react-shared';
import CodeTab from './Tabs/Code/CodeTab';
import LambdaCode from './Tabs/Code/LambdaCode';
import LambdaDependencies from './Tabs/Code/LambdaDependencies';
import ConfigurationTab from './Tabs/Configuration/Configuration';

const exampleLambdaCode = `module.exports = { 
  main: function (event, context) {

  }
}`;

const exampleDependencies = `{ 
  "name": "%NAME%",
  "version": "1.0.0",
  "dependencies": {}
}`;

export default function LambdaDetails({ lambda }) {
  const [labels, setLabels] = useState(lambda.labels);
  const [lambdaCode, setLambdaCode] = useState(
    lambda.content || exampleLambdaCode,
  );
  const [dependencies, setDependencies] = useState(
    lambda.dependencies || exampleDependencies.replace('%NAME%', lambda.name),
  );
  const [selectedTabIndex, setSelectedTabIndex] = useState(0);
  const [updateLambdaMutation] = useMutation(UPDATE_LAMBDA);
  const notificationManager = useNotification();

  const formRef = useRef(null);

  const formValues = {
    size: useRef(null),
    runtime: useRef(null),
  };

  const namespace = LuigiClient.getEventData().environmentId;
  const selectedTabName =
    LuigiClient.getNodeParams().selectedTab || 'Configuration';
  useEffect(() => {
    const selectedTabIndex = selectedTabName === 'Configuration' ? 0 : 1;
    setSelectedTabIndex(selectedTabIndex);
  }, [selectedTabName]);

  async function updateLambda() {
    if (!formRef.current.checkValidity()) {
      return;
    }

    try {
      const updatedFunction = await updateLambdaMutation({
        variables: {
          name: lambda.name,
          namespace,
          params: {
            labels: labels || {},
            size: formValues.size.current.value,
            runtime: formValues.runtime.current.value,
            content: lambdaCode,
            dependencies,
          },
        },
      });

      const isSuccess =
        updatedFunction.data &&
        updatedFunction.data.updateFunction &&
        updatedFunction.data.updateFunction.name === lambda.name;
      if (isSuccess) {
        notificationManager.notify({
          content: `Lambda ${lambda.name} updated successfully`,
          title: 'Success',
          color: '#107E3E',
          icon: 'accept',
          autoClose: true,
        });
      }
    } catch (e) {
      notificationManager.notify({
        content: `Error while removing lambda ${lambda.name}: ${e.message}`,
        title: 'Error',
        color: '#BB0000',
        icon: 'decline',
        autoClose: false,
      });
    }
  }

  function updateLabels(newLabels) {
    setLabels(newLabels);
  }

  const onChangeTab = (_, id) => {
    const selectedTab = id === 0 ? 'Configuration' : 'Code';
    try {
      LuigiClient.linkManager()
        .withParams({ selectedTab })
        .navigate('');
    } catch (e) {
      console.error(e);
    }
  };

  return (
    <>
      <LambdaDetailsHeader
        lambda={lambda}
        handleUpdate={updateLambda}
      ></LambdaDetailsHeader>
      <TabGroup selectedIndex={selectedTabIndex} onTabClick={onChangeTab}>
        <Tab
          key="lambda-configuration"
          id="lambda-configuration"
          title="Configuration"
        >
          <ConfigurationTab
            lambda={lambda}
            formRef={formRef}
            sizeRef={formValues.size}
            runtimeRef={formValues.runtime}
            LabelsEditor={
              <LabelSelectorInput labels={labels} onChange={updateLabels} />
            }
          />
        </Tab>

        <Tab key="lambda-code" id="lambda-code" title="Code">
          <CodeTab
            codeEditorComponent={
              <LambdaCode code={lambdaCode} setCode={setLambdaCode} />
            }
            dependenciesComponent={
              <LambdaDependencies
                dependencies={dependencies}
                setDependencies={setDependencies}
              />
            }
          />
        </Tab>
      </TabGroup>
    </>
  );
}

LambdaDetails.propTypes = {
  lambda: PropTypes.object.isRequired,
};
