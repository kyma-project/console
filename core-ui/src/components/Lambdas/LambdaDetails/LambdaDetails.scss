.lambda-details {
  position: relative;
  height: 100vh;

  > .fd-loading-dots {
    margin: 0 auto;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  .lambda-details-tabs {
    .fd-tabs__item:first-of-type {
      padding-left: 32px;
    }
  }
}