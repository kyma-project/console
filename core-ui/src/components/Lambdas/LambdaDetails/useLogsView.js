import { useEffect, useState } from 'react';
import LuigiClient from '@kyma-project/luigi-client';

export const useLogsView = (name, namespace, selectedTabName) => {
  const [logsViewExists, setLogViewExists] = useState(false);

  useEffect(() => {
    const linkManager = LuigiClient.linkManager().withParams({
      namespace: namespace,
      instance: `~${name}-.+`,
      compact: 'true',
    });

    linkManager
      .pathExists('/home/cmf-logs')
      .then(exists => setLogViewExists(exists));

    let logsViewHandle;

    if (logsViewExists) {
      logsViewHandle = linkManager.openAsSplitView('/home/cmf-logs', {
        title: 'Logs',
        size: 40,
        collapsed: true,
      });

      if (selectedTabName === 'Configuration') {
        logsViewHandle.collapse();
      } else {
        logsViewHandle.expand();
      }
    }

    return () => {
      !!logsViewHandle && logsViewHandle.collapse();
    };
  }, [namespace, name, selectedTabName, logsViewExists]);
};
