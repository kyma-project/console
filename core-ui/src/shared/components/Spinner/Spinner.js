import React from 'react';

const Spinner = () => {
  return (
    <div className="fd-loading-dots" aria-hidden="false" aria-label="Loading">
      <div></div>
      <div></div>
      <div></div>
    </div>
  );
};

export default Spinner;
