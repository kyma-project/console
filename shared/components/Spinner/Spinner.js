import React from 'react';

export const Spinner = () => {
  return (
    <div className="fd-loading-dots" aria-hidden="false" aria-label="Loading">
      <div></div>
      <div></div>
      <div></div>
    </div>
  );
};
