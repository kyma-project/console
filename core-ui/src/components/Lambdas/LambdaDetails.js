import React from 'react';

import LambdaDetailsWrapper from 'components/Lambdas/LambdaDetails/LambdaDetailsWrapper';
import { useConfigData } from 'components/Lambdas/helpers/misc/useConfigData';

export default function LambdaDetails({ match }) {
  useConfigData();
  return <LambdaDetailsWrapper lambdaName={match.params.name} />;
}
