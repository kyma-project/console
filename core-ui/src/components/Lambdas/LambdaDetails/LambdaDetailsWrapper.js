import React from 'react';
import LuigiClient from '@kyma-project/luigi-client';
import PropTypes from 'prop-types';

import { useQuery } from '@apollo/react-hooks';
import { GET_LAMBDA } from '../../../gql/queries';
import { REFETCH_TIMEOUT } from '../../../shared/constants';

import EntryNotFound from '../../EntryNotFound/EntryNotFound';
import { Spinner } from 'react-shared';
import LambdaDetails from './LambdaDetails';

LambdaDetailsWrapper.propTypes = {
  lambdaName: PropTypes.string.isRequired,
};

export default function LambdaDetailsWrapper({ lambdaName }) {
  const namespace = LuigiClient.getEventData().environmentId;
  const { data, error, loading, refetch } = useQuery(GET_LAMBDA, {
    variables: {
      name: lambdaName,
      namespace,
    },
    fetchPolicy: 'no-cache',
  });

  if (error) {
    return `Error! ${error.message}`;
  }
  if (loading) {
    return <Spinner />;
  }
  if (data && !data.function) {
    setTimeout(() => {
      refetch();
    }, REFETCH_TIMEOUT);
    return <EntryNotFound entryType="Lambda" entryId={lambdaName} />;
  }
  if (data && data.function) {
    return <LambdaDetails lambda={data.function} />;
  }
}
