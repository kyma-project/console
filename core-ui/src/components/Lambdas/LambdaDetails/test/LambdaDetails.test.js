import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';

import {
  getLambdaSuccessMock,
  getLambdaNoContentSuccessMock,
  lambda,
} from './gqlMocks';
import { componentUpdate } from '../../../../testing';
import LambdaDetails from '../LambdaDetails';
import Spinner from '../../../../shared/components/Spinner/Spinner';

jest.mock('@kyma-project/luigi-client', () => {
  return {
    getEventData: () => ({ environmentId: 'testnamespace' }),
    getNodeParams: () => ({ selectedTab: 'Code' }),
  };
});

describe('LambdaDetails', () => {
  it('Renders with an error in no lambdaId is provided', async () => {
    console.error = jest.fn();
    let component;
    await renderer.act(async () => {
      component = renderer.create(
        <MockedProvider>
          <LambdaDetails />
        </MockedProvider>,
      );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(console.error.mock.calls.length).toBe(1);
    expect(console.error.mock.calls[0][0]).toMatchSnapshot();
  });

  it('Shows loading indicator only when data is not yet loaded', async () => {
    const component = mount(
      <MockedProvider>
        <LambdaDetails lambdaId={lambda.name} />
      </MockedProvider>,
    );
    expect(component.find(Spinner)).toHaveLength(1);

    await componentUpdate(component);

    expect(component.find(Spinner)).toHaveLength(0);
  });

  it('Displays lambda, its size, runtime and content', async () => {
    const gqlMock = [getLambdaSuccessMock()];

    const component = mount(
      <MockedProvider mocks={gqlMock}>
        <LambdaDetails lambdaId={lambda.name} />
      </MockedProvider>,
    );

    await componentUpdate(component);

    const lambdaSizeInput = component.find('#lambdaSize');
    expect(lambdaSizeInput.exists()).toEqual(true);
    expect(lambdaSizeInput.props().defaultValue).toEqual(lambda.size);

    const lambdaRuntimeInput = component.find('#lambdaRuntime');
    expect(lambdaRuntimeInput.exists()).toEqual(true);
    expect(lambdaRuntimeInput.props().defaultValue).toEqual(lambda.runtime);

    const lambdaContentInput = component.find('#lambdaContent');
    expect(lambdaContentInput.exists()).toEqual(true);
    expect(lambdaContentInput.props().value).toEqual(lambda.content);
  });

  it('Displays lambda, its size, runtime and default content', async () => {
    const gqlMock = [getLambdaNoContentSuccessMock()];

    const component = mount(
      <MockedProvider mocks={gqlMock}>
        <LambdaDetails lambdaId={lambda.name} />
      </MockedProvider>,
    );

    await componentUpdate(component);

    const lambdaSizeInput = component.find('#lambdaSize');
    expect(lambdaSizeInput.exists()).toEqual(true);
    expect(lambdaSizeInput.props().defaultValue).toEqual(lambda.size);

    const lambdaRuntimeInput = component.find('#lambdaRuntime');
    expect(lambdaRuntimeInput.exists()).toEqual(true);
    expect(lambdaRuntimeInput.props().defaultValue).toEqual(lambda.runtime);

    const lambdaContentInput = component.find('#lambdaContent');
    expect(lambdaContentInput.exists()).toEqual(true);
    expect(lambdaContentInput.props().value).toEqual(
      "console.log('Hello World!');",
    );
  });
});
