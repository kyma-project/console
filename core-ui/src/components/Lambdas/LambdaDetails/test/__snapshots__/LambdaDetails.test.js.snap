// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaDetails Renders with an error in no lambdaId is provided 1`] = `
"Error! Network error: No more mocked responses for the query: query Function($name: String!, $namespace: String!) {
  function(name: $name, namespace: $namespace) {
    name
    namespace
    labels
    runtime
    size
    status
    content
    dependencies
    __typename
  }
}
, variables: {\\"namespace\\":\\"testnamespace\\"}"
`;

exports[`LambdaDetails Renders with an error in no lambdaId is provided 2`] = `
"Warning: Failed prop type: The prop \`lambdaId\` is marked as required in \`LambdaDetails\`, but its value is \`undefined\`.
    in LambdaDetails (at LambdaDetails.test.js:29)"
`;
